File: dart/lib/src/sse.dart
import 'dart:convert';

/// Parse a Server-Sent Events byte stream into a stream of decoded JSON values.
/// Iteration stops at the `[DONE]` sentinel.
Stream<dynamic> parseSse(Stream<List<int>> byteStream) async* {
  final lines = byteStream.transform(utf8.decoder).transform(const LineSplitter());
  final data = <String>[];

  await for (final line in lines) {
    if (line.isEmpty) {
      if (data.isEmpty) continue;
      final payload = data.join('\n');
      data.clear();
      if (payload == '[DONE]') return;
      yield jsonDecode(payload);
    } else if (line.startsWith(':')) {
      // comment / keep-alive
    } else if (line.startsWith('data:')) {
      var value = line.substring(5);
      if (value.startsWith(' ')) value = value.substring(1);
      data.add(value);
    }
  }

  if (data.isNotEmpty) {
    final payload = data.join('\n');
    if (payload != '[DONE]') yield jsonDecode(payload);
  }
}

File: rust/src/sse.rs
//! Server-Sent Events parsing into a stream of JSON values.

use crate::error::{Error, Result};
use futures::Stream;
use futures::StreamExt;
use serde_json::Value;

/// Turn an SSE HTTP response into a stream of parsed JSON events. Iteration stops
/// at the `[DONE]` sentinel.
pub(crate) fn parse_sse(resp: reqwest::Response) -> impl Stream<Item = Result<Value>> {
    async_stream::try_stream! {
        let mut bytes = resp.bytes_stream();
        let mut buf = String::new();
        let mut data: Vec<String> = Vec::new();

        while let Some(chunk) = bytes.next().await {
            let chunk = chunk.map_err(|e| Error::Connection(e.to_string()))?;
            buf.push_str(&String::from_utf8_lossy(&chunk));
            buf = buf.replace("\r\n", "\n").replace('\r', "\n");

            while let Some(idx) = buf.find('\n') {
                let line: String = buf.drain(..=idx).collect();
                let line = line.trim_end_matches('\n');

                if line.is_empty() {
                    if data.is_empty() {
                        continue;
                    }
                    let payload = std::mem::take(&mut data).join("\n");
                    if payload == "[DONE]" {
                        return;
                    }
                    let value: Value = serde_json::from_str(&payload)?;
                    yield value;
                } else if line.starts_with(':') {
                    // comment / keep-alive
                } else if let Some(rest) = line.strip_prefix("data:") {
                    data.push(rest.strip_prefix(' ').unwrap_or(rest).to_string());
                }
            }
        }

        if !data.is_empty() {
            let payload = data.join("\n");
            if payload != "[DONE]" {
                let value: Value = serde_json::from_str(&payload)?;
                yield value;
            }
        }
    }
}
